File: backend/app/utils/formatting.py
from datetime import datetime
from typing import Optional, Union
import dateutil.parser


def format_kg(weight_gram_or_kg: Optional[Union[int, float]], from_gram: bool = True) -> str:
    """Format weight in kilogram with 3 decimal precision (e.g. 2,500 kg)."""
    if weight_gram_or_kg is None:
        return "0,000 kg"
    if from_gram:
        kg = weight_gram_or_kg / 1000.0
    else:
        kg = float(weight_gram_or_kg)
    formatted = f"{kg:,.3f}".replace(",", "X").replace(".", ",").replace("X", ".")
    return f"{formatted} kg"


def format_date(dt: Optional[Union[str, datetime]]) -> str:
    """Format date to DD/MM/YYYY."""
    if not dt:
        return "-"
    if isinstance(dt, str):
        try:
            parsed = dateutil.parser.parse(dt)
        except Exception:
            return dt
    else:
        parsed = dt
    return parsed.strftime("%d/%m/%Y")


def format_datetime(dt: Optional[Union[str, datetime]]) -> str:
    """Format datetime to DD/MM/YYYY HH:MM."""
    if not dt:
        return "-"
    if isinstance(dt, str):
        try:
            parsed = dateutil.parser.parse(dt)
        except Exception:
            return dt
    else:
        parsed = dt
    return parsed.strftime("%d/%m/%Y %H:%M")

File: backend/src/utils/formatting.ts
export function formatKg(
  weightGramOrKg: number | null | undefined,
  fromGram: boolean = true
): string {
  if (weightGramOrKg === null || weightGramOrKg === undefined) {
    return '0,000 kg';
  }
  const kg = fromGram ? Number(weightGramOrKg) / 1000.0 : Number(weightGramOrKg);
  // Format with 3 decimals, using comma as decimal separator and dot as thousand separator
  const parts = kg.toFixed(3).split('.');
  const integerPart = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, '.');
  const decimalPart = parts[1];
  return `${integerPart},${decimalPart} kg`;
}

export function formatDate(dt: string | Date | null | undefined): string {
  if (!dt) return '-';
  const d = typeof dt === 'string' ? new Date(dt) : dt;
  if (isNaN(d.getTime())) return typeof dt === 'string' ? dt : '-';
  const day = String(d.getDate()).padStart(2, '0');
  const month = String(d.getMonth() + 1).padStart(2, '0');
  const year = d.getFullYear();
  return `${day}/${month}/${year}`;
}

export function formatDateTime(dt: string | Date | null | undefined): string {
  if (!dt) return '-';
  const d = typeof dt === 'string' ? new Date(dt) : dt;
  if (isNaN(d.getTime())) return typeof dt === 'string' ? dt : '-';
  const day = String(d.getDate()).padStart(2, '0');
  const month = String(d.getMonth() + 1).padStart(2, '0');
  const year = d.getFullYear();
  const hours = String(d.getHours()).padStart(2, '0');
  const minutes = String(d.getMinutes()).padStart(2, '0');
  return `${day}/${month}/${year} ${hours}:${minutes}`;
}
